cpp/include/raft/cluster/kmeans_balanced.cuh

```diff
@@ -0,0 +1,365 @@
+ *
+ * This variant of the k-means algorithm first clusters the dataset in mesoclusters, then clusters
+ * the subsets associated to each mesocluster into fine clusters, and finally runs a few k-means
+ * iterations over the whole dataset and with all the centroids to obtain the final clusters.
+ *
+ * Each k-means iteration applies expectation-maximization-balancing:
+ *  - Balancing: adjust centers for clusters that have a small number of entries. If the size of a
+ *    cluster is below a threshold, the center is moved towards a bigger cluster.
+ *  - Expectation: predict the labels (i.e find closest cluster centroid to each point)
+ *  - Maximization: calculate optimal centroids (i.e find the center of gravity of each cluster)
+ *
+ * The number of mesoclusters is chosen by rounding the square root of the number of clusters. E.g
+ * for 512 clusters, we would have 23 mesoclusters. The number of fine clusters per mesocluster is
+ * chosen proportionally to the number of points in each mesocluster.
+ *
+ * This variant of k-means uses random initialization and a fixed number of iterations, though
+ * iterations can be repeated if the balancing step moved the centroids.
+ *
+ * Additionally, this algorithm supports quantized datasets in arbitrary types but the core part of
+ * the algorithm will work with a floating-point type, hence a conversion function can be provided
+ * to map the data type to the math type.
+ *
+ * @code{.cpp}
+ *   #include <raft/core/handle.hpp>
+ *   #include <raft/cluster/kmeans_balanced.cuh>
+ *   #include <raft/cluster/kmeans_balanced_types.hpp>
+ *   ...
+ *   raft::handle_t handle;
+ *   raft::cluster::kmeans_balanced_params params;
+ *   auto centroids = raft::make_device_matrix<float, int>(handle, n_clusters, n_features);
+ *   raft::cluster::kmeans_balanced::fit(handle, params, X, centroids.view());
+ * @endcode
+ *
+ * @tparam DataT Type of the input data.
+ * @tparam MathT Type of the centroids and mapped data.
+ * @tparam IndexT Type used for indexing.
+ * @tparam MappingOpT Type of the mapping function.
+ * @param[in]  handle     The raft resources
+ * @param[in]  params     Structure containing the hyper-parameters
+ * @param[in]  X          Training instances to cluster. The data must be in row-major format.
+ *                        [dim = n_samples x n_features]
+ * @param[out] centroids  The generated centroids [dim = n_clusters x n_features]
+ * @param[in]  mapping_op (optional) Functor to convert from the input datatype to the arithmetic
+ *                        datatype. If DataT == MathT, this must be the identity.
+ */
+template <typename DataT, typename MathT, typename IndexT, typename MappingOpT = raft::identity_op>
+void fit(const raft::device_resources& handle,
+         kmeans_balanced_params const& params,
+         raft::device_matrix_view<const DataT, IndexT> X,
+         raft::device_matrix_view<MathT, IndexT> centroids,
+         MappingOpT mapping_op = raft::identity_op())
+{
```

Comment:
If you require the mapping to be identity when `DataT == MathT`, consider adding a static assert:
```C++
static_assert(!std::is_same_v<MathT, DataT> || !std::is_same_v<MappingOpT, raft::identity_op>,
  "Only the identity mapping is supported when MathT == DataT")
```

Reply:
No, this would break the ANN methods that pass `mapping<float>` which behaves like the identity when the input type is float. We want *an* identity op, not *the* `identity_op`.